File: ios/iOSBaseProject/Network/ClerkAuthService.swift
//
//  ClerkAuthService.swift
//  BaatCheet
//
//  Authentication Service for BaatCheet - Uses Backend API
//

import Foundation

// MARK: - API Configuration
struct APIConfig {
    // Backend API URL - adjust for your environment
    // Use your computer's local IP for physical device testing
    #if DEBUG
    static let baseURL = "http://192.168.18.110:5001/api/v1"
    #else
    static let baseURL = "https://api.baatcheet.com/api/v1"
    #endif
    
    static let mobileAuthURL = "\(baseURL)/mobile/auth"
}

// MARK: - Auth Result
enum ClerkAuthResult {
    case success(token: String, userId: String)
    case needsVerification
    case failure(Error)
}

// MARK: - API Response Types
struct AuthResponse: Codable {
    let success: Bool
    let data: AuthData?
    let error: String?
}

struct AuthData: Codable {
    let user: UserData?
    let token: String?
    let status: String?
    let message: String?
    let email: String?
}

struct UserData: Codable {
    let id: String
    let email: String
    let firstName: String?
    let lastName: String?
    let avatar: String?
    let role: String?
    let tier: String?
}

// MARK: - ClerkAuthService
class ClerkAuthService: ObservableObject {
    static let shared = ClerkAuthService()
    
    @Published var isAuthenticated = false
    @Published var currentUserId: String?
    @Published var currentUser: UserData?
    
    private let tokenKey = "baatcheet_auth_token"
    private let userKey = "baatcheet_user_data"
    private var pendingEmail: String?

    private init() {
        // Check for existing session on init
        if let token = getAuthToken() {
            isAuthenticated = true
            loadUserFromStorage()
            // Optionally verify token with server
            Task {
                await refreshUserInfo()
            }
        }
    }
    
    // MARK: - Sign Up with Email
    func signUp(email: String, password: String, firstName: String? = nil, lastName: String? = nil) async -> ClerkAuthResult {
        guard let url = URL(string: "\(APIConfig.mobileAuthURL)/signup") else {
            return .failure(AppError.apiError("Invalid URL"))
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        var body: [String: Any] = [
            "email": email,
            "password": password
        ]
        if let firstName = firstName {
            body["firstName"] = firstName
        }
        if let lastName = lastName {
            body["lastName"] = lastName
        }
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(AppError.apiError("Invalid response"))
            }
            
            let decoder = JSONDecoder()
            let authResponse = try decoder.decode(AuthResponse.self, from: data)
            
            if authResponse.success {
                if let authData = authResponse.data {
                    // Check if verification is needed
                    if authData.status == "verification_required" {
                        self.pendingEmail = email
                        return .needsVerification
                    }
                    
                    // User created successfully
                    if let token = authData.token, let user = authData.user {
                        await MainActor.run {
                            self.saveAuthToken(token)
                            self.saveUser(user)
                            self.isAuthenticated = true
                            self.currentUserId = user.id
                            self.currentUser = user
                        }
                        return .success(token: token, userId: user.id)
                    }
                }
            }
            
            // Error response
            let errorMessage = authResponse.error ?? "Sign up failed"
            return .failure(AppError.apiError(errorMessage))
            
        } catch {
            print("Sign up error: \(error.localizedDescription)")
            return .failure(error)
        }
    }
    
    // MARK: - Sign In with Email
    func signIn(email: String, password: String) async -> ClerkAuthResult {
        guard let url = URL(string: "\(APIConfig.mobileAuthURL)/signin") else {
            return .failure(AppError.apiError("Invalid URL"))
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "email": email,
            "password": password
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                return .failure(AppError.apiError("Invalid response"))
            }
            
            let decoder = JSONDecoder()
            let authResponse = try decoder.decode(AuthResponse.self, from: data)
            
            if authResponse.success {
                if let authData = authResponse.data,
                   let token = authData.token,
                   let user = authData.user {
                    await MainActor.run {
                        self.saveAuthToken(token)
                        self.saveUser(user)
                        self.isAuthenticated = true
                        self.currentUserId = user.id
                        self.currentUser = user
                    }
                    return .success(token: token, userId: user.id)
                }
            }
            
            // Check for specific error
            let errorMessage = authResponse.error ?? "Sign in failed"
            
            // If user not found, they need to sign up
            if httpResponse.statusCode == 401 {
                return .failure(AppError.apiError("Invalid email or password. Please check your credentials or sign up."))
            }
            
            return .failure(AppError.apiError(errorMessage))
            
        } catch {
            print("Sign in error: \(error.localizedDescription)")
            return .failure(error)
        }
    }
    
    // MARK: - Verify Email Code
    func verifyEmailCode(code: String) async -> ClerkAuthResult {
        guard let email = pendingEmail else {
            return .failure(AppError.apiError("No pending verification. Please sign up again."))
        }
        
        guard let url = URL(string: "\(APIConfig.mobileAuthURL)/verify-email") else {
            return .failure(AppError.apiError("Invalid URL"))
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "email": email,
            "code": code
        ]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            
            let (data, response) = try await URLSession.shared.data(for: request)
            
            let decoder = JSONDecoder()
            let authResponse = try decoder.decode(AuthResponse.self, from: data)
            
            if authResponse.success {
                if let authData = authResponse.data,
                   let token = authData.token,
                   let user = authData.user {
                    await MainActor.run {
                        self.saveAuthToken(token)
                        self.saveUser(user)
                        self.isAuthenticated = true
                        self.currentUserId = user.id
                        self.currentUser = user
                        self.pendingEmail = nil
                    }
                    return .success(token: token, userId: user.id)
                }
            }
            
            let errorMessage = authResponse.error ?? "Verification failed"
            return .failure(AppError.apiError(errorMessage))
            
        } catch {
            print("Verify email error: \(error.localizedDescription)")
            return .failure(error)
        }
    }
    
    // MARK: - Resend Verification Code
    func resendVerificationCode() async -> Bool {
        guard let email = pendingEmail else {
            print("No pending email for resend")
            return false
        }
        
        guard let url = URL(string: "\(APIConfig.mobileAuthURL)/resend-code") else {
            return false
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = ["email": email]
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                return false
            }
            
            return httpResponse.statusCode == 200
        } catch {
            print("Resend code error: \(error.localizedDescription)")
            return false
        }
    }
    
    // MARK: - Refresh User Info
    func refreshUserInfo() async {
        guard let token = getAuthToken() else { return }
        
        guard let url = URL(string: "\(APIConfig.mobileAuthURL)/me") else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                // Token might be invalid
                if (response as? HTTPURLResponse)?.statusCode == 401 {
                    await MainActor.run {
                        self.clearSession()
                    }
                }
                return
            }
            
            struct MeResponse: Codable {
                let success: Bool
                let data: MeData?
            }
            struct MeData: Codable {
                let user: UserData
            }
            
            let decoder = JSONDecoder()
            let meResponse = try decoder.decode(MeResponse.self, from: data)
            
            if let user = meResponse.data?.user {
                await MainActor.run {
                    self.saveUser(user)
                    self.currentUser = user
                    self.currentUserId = user.id
                }
            }
        } catch {
            print("Refresh user info error: \(error.localizedDescription)")
        }
    }
    
    // MARK: - Token Management
    func saveAuthToken(_ token: String) {
        UserDefaults.standard.set(token, forKey: tokenKey)
    }
    
    func getAuthToken() -> String? {
        return UserDefaults.standard.string(forKey: tokenKey)
    }
    
    private func saveUser(_ user: UserData) {
        if let encoded = try? JSONEncoder().encode(user) {
            UserDefaults.standard.set(encoded, forKey: userKey)
        }
    }
    
    private func loadUserFromStorage() {
        if let data = UserDefaults.standard.data(forKey: userKey),
           let user = try? JSONDecoder().decode(UserData.self, from: data) {
            self.currentUser = user
            self.currentUserId = user.id
        }
    }
    
    func clearSession() {
        UserDefaults.standard.removeObject(forKey: tokenKey)
        UserDefaults.standard.removeObject(forKey: userKey)
        isAuthenticated = false
        currentUserId = nil
        currentUser = nil
        pendingEmail = nil
    }
    
    // MARK: - Sign Out
    @MainActor
    func signOut() {
        Task {
            // Call logout endpoint
            if let token = getAuthToken(),
               let url = URL(string: "\(APIConfig.mobileAuthURL)/logout") {
                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                
                _ = try? await URLSession.shared.data(for: request)
            }
        }
        
        clearSession()
        AuthenticationManager.shared.logout()
    }
    
    // MARK: - Set Pending Email (for verification flow)
    func setPendingEmail(_ email: String) {
        self.pendingEmail = email
    }
    
    func getPendingEmail() -> String? {
        return pendingEmail
    }
}

File: android/app/src/main/java/com/baatcheet/app/ui/login/ClerkAuthService.kt
package com.baatcheet.app.ui.login

import android.content.Context
import android.content.SharedPreferences
import androidx.lifecycle.ViewModel
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import javax.inject.Inject

// MARK: - API Configuration
object APIConfig {
    // Backend API URL - HuggingFace Spaces deployment
    const val BASE_URL = "https://sharry121-baatcheet.hf.space/api/v1"
    const val MOBILE_AUTH_URL = "$BASE_URL/mobile/auth"
}

// MARK: - Response Models
data class AuthResponse(
    val success: Boolean,
    val data: AuthData?,
    val error: String?
)

data class AuthData(
    val user: UserData?,
    val token: String?,
    val status: String?,
    val message: String?,
    val email: String?
)

data class UserData(
    val id: String,
    val email: String,
    @SerializedName("firstName") val firstName: String?,
    @SerializedName("lastName") val lastName: String?,
    val avatar: String?,
    val role: String?,
    val tier: String?
)

// MARK: - Auth Result
sealed class ClerkAuthResult {
    data class Success(val token: String, val userId: String, val user: UserData? = null) : ClerkAuthResult()
    object NeedsVerification : ClerkAuthResult()
    data class Failure(val error: Throwable) : ClerkAuthResult()
}

// MARK: - ClerkAuthService
@HiltViewModel
class ClerkAuthService @Inject constructor(
    @ApplicationContext private val context: Context
) : ViewModel() {
    
    private val okHttpClient = OkHttpClient.Builder().build()
    private val gson = Gson()
    private val prefs: SharedPreferences = context.getSharedPreferences("baatcheet_auth", Context.MODE_PRIVATE)
    
    private var pendingEmail: String? = null
    
    companion object {
        private const val TOKEN_KEY = "auth_token"
        private const val USER_KEY = "user_data"
    }

    // MARK: - Sign Up with Email
    suspend fun signUp(email: String, password: String, firstName: String? = null, lastName: String? = null): ClerkAuthResult = withContext(Dispatchers.IO) {
        val url = "${APIConfig.MOBILE_AUTH_URL}/signup"
        
        val body = mutableMapOf(
            "email" to email,
            "password" to password
        )
        firstName?.let { body["firstName"] = it }
        lastName?.let { body["lastName"] = it }
        
        val requestBody = gson.toJson(body).toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .build()

        try {
            val response = okHttpClient.newCall(request).execute()
            val responseBody = response.body?.string()
            
            if (responseBody != null) {
                val authResponse = gson.fromJson(responseBody, AuthResponse::class.java)
                
                if (authResponse.success) {
                    authResponse.data?.let { data ->
                        // Check if verification is needed
                        if (data.status == "verification_required") {
                            pendingEmail = email
                            return@withContext ClerkAuthResult.NeedsVerification
                        }
                        
                        // Check if web signup is required (legacy - should not happen anymore)
                        if (data.status == "web_signup_required") {
                            return@withContext ClerkAuthResult.Failure(
                                Exception(data.message ?: "Please sign up via web first")
                            )
                        }
                        
                        // User created successfully
                        if (data.token != null && data.user != null) {
                            saveAuthToken(data.token)
                            saveUser(data.user)
                            return@withContext ClerkAuthResult.Success(
                                token = data.token,
                                userId = data.user.id,
                                user = data.user
                            )
                        }
                        
                        // If we have a message but no token, show the message
                        if (data.message != null && data.token == null) {
                            return@withContext ClerkAuthResult.Failure(Exception(data.message))
                        }
                    }
                }
                
                // Error response
                val errorMessage = authResponse.error ?: "Sign up failed"
                return@withContext ClerkAuthResult.Failure(Exception(errorMessage))
            }
            
            ClerkAuthResult.Failure(Exception("Empty response from server"))
        } catch (e: IOException) {
            ClerkAuthResult.Failure(e)
        } catch (e: Exception) {
            ClerkAuthResult.Failure(e)
        }
    }

    // MARK: - Sign In with Email
    suspend fun signIn(email: String, password: String): ClerkAuthResult = withContext(Dispatchers.IO) {
        val url = "${APIConfig.MOBILE_AUTH_URL}/signin"
        
        val body = mapOf(
            "email" to email,
            "password" to password
        )
        val requestBody = gson.toJson(body).toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .build()

        try {
            val response = okHttpClient.newCall(request).execute()
            val responseBody = response.body?.string()
            
            if (responseBody != null) {
                val authResponse = gson.fromJson(responseBody, AuthResponse::class.java)
                
                if (authResponse.success) {
                    authResponse.data?.let { data ->
                        if (data.token != null && data.user != null) {
                            saveAuthToken(data.token)
                            saveUser(data.user)
                            return@withContext ClerkAuthResult.Success(
                                token = data.token,
                                userId = data.user.id,
                                user = data.user
                            )
                        }
                    }
                }
                
                // Error response
                val errorMessage = authResponse.error ?: "Sign in failed"
                return@withContext ClerkAuthResult.Failure(Exception(errorMessage))
            }
            
            ClerkAuthResult.Failure(Exception("Empty response from server"))
        } catch (e: IOException) {
            ClerkAuthResult.Failure(e)
        } catch (e: Exception) {
            ClerkAuthResult.Failure(e)
        }
    }

    // MARK: - Verify Email Code
    suspend fun verifyEmailCode(code: String): ClerkAuthResult = withContext(Dispatchers.IO) {
        val email = pendingEmail ?: return@withContext ClerkAuthResult.Failure(
            Exception("No pending verification. Please sign up again.")
        )
        
        val url = "${APIConfig.MOBILE_AUTH_URL}/verify-email"
        
        val body = mapOf(
            "email" to email,
            "code" to code
        )
        val requestBody = gson.toJson(body).toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .build()

        try {
            val response = okHttpClient.newCall(request).execute()
            val responseBody = response.body?.string()
            
            if (responseBody != null) {
                val authResponse = gson.fromJson(responseBody, AuthResponse::class.java)
                
                if (authResponse.success) {
                    authResponse.data?.let { data ->
                        if (data.token != null && data.user != null) {
                            saveAuthToken(data.token)
                            saveUser(data.user)
                            pendingEmail = null
                            return@withContext ClerkAuthResult.Success(
                                token = data.token,
                                userId = data.user.id,
                                user = data.user
                            )
                        }
                    }
                }
                
                // Error response
                val errorMessage = authResponse.error ?: "Verification failed"
                return@withContext ClerkAuthResult.Failure(Exception(errorMessage))
            }
            
            ClerkAuthResult.Failure(Exception("Empty response from server"))
        } catch (e: IOException) {
            ClerkAuthResult.Failure(e)
        } catch (e: Exception) {
            ClerkAuthResult.Failure(e)
        }
    }

    // MARK: - Resend Verification Code
    suspend fun resendVerificationCode(): Boolean = withContext(Dispatchers.IO) {
        val email = pendingEmail ?: return@withContext false
        
        val url = "${APIConfig.MOBILE_AUTH_URL}/resend-code"
        
        val body = mapOf("email" to email)
        val requestBody = gson.toJson(body).toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .header("Content-Type", "application/json")
            .build()

        try {
            val response = okHttpClient.newCall(request).execute()
            response.isSuccessful
        } catch (e: Exception) {
            false
        }
    }

    // MARK: - Token Management
    private fun saveAuthToken(token: String) {
        prefs.edit().putString(TOKEN_KEY, token).apply()
    }

    fun getAuthToken(): String? {
        return prefs.getString(TOKEN_KEY, null)
    }

    private fun saveUser(user: UserData) {
        val userJson = gson.toJson(user)
        prefs.edit().putString(USER_KEY, userJson).apply()
    }

    fun getUser(): UserData? {
        val userJson = prefs.getString(USER_KEY, null)
        return if (userJson != null) {
            try {
                gson.fromJson(userJson, UserData::class.java)
            } catch (e: Exception) {
                null
            }
        } else {
            null
        }
    }

    fun clearSession() {
        prefs.edit().clear().apply()
        pendingEmail = null
    }

    // MARK: - Check Authentication
    fun isAuthenticated(): Boolean {
        return getAuthToken() != null
    }

    // MARK: - Set Pending Email
    fun setPendingEmail(email: String) {
        pendingEmail = email
    }

    fun getPendingEmail(): String? {
        return pendingEmail
    }

    // MARK: - Logout
    suspend fun logout(): Boolean = withContext(Dispatchers.IO) {
        val token = getAuthToken()
        if (token != null) {
            try {
                val url = "${APIConfig.MOBILE_AUTH_URL}/logout"
                val request = Request.Builder()
                    .url(url)
                    .post("{}".toRequestBody("application/json".toMediaType()))
                    .header("Authorization", "Bearer $token")
                    .build()
                
                okHttpClient.newCall(request).execute()
            } catch (e: Exception) {
                // Ignore logout errors
            }
        }
        clearSession()
        true
    }
}
